Implement recommended resolution based on available outgoing bitrate

vona-ben opened the pull request:
Publisher statistics and availableOutgoingBitrate provide more accurate estimation, especially if simulcast is enabled.

Changes made:

- Use availableOutgoingBitrate to determine the resolution. Remove the threshold for packet loss (availableOutgoingBitrate will be automatically adjusted with network conditions).
- Expose bytesSent statistics for both video and audio and display bytesSent instead of bytesReceived for the sample.

Note: Since availableOutgoingBitrate is not available on Firefox, I have removed support for it in this browser.

Fixes included

- Detect when the publisher leaves the session. Once the publisher is detected as having left the session, stop sending stats.
- RoundTripTime was never defined previously, fixed used new publisher stats interface
- Use session.unpublish instead of destroy()

rhainer reviewed (commented):
Why is this PR being submitted to master branch. I suggest we test this from develop branch first.

vona-ben commented:
@jeffswartz, I removed availableOutgoingBitrate from the public API. I am still using the RTC internal interface because I have also added a few helpers. It will take some time for me to transition to using the native interface

andresbersier commented on src/NetworkTest/testQuality/helpers/config.ts:
```diff
@@ -38,34 +38,28 @@ const config: QualityTestConfig = {
-        bps: 1000000,
-        plr: 0.005,
-        recommendedSetting: '1280x720 @ 30FPS',
+        bps: 4000000,
```
We risk being too conservative by using these values, since the browser will actually start enoding at FHD (for vp8) with much less available bitrate than the target bitrate, which is codec independent. See https://jira.vonage.com/browse/VIDMR-73

vona-ben replied:
Hey @andresbersier, we are currently using the ongoing available bitrate (I have renamed the variable). Based on my observations, it seems that whenever the outgoingAvailableBitrate reaches the target bitrate, the browser initiates a resolution change or adjusts the codec, regardless of the specific bitrate we are sending. I have noticed clear resolution changes during these values in my tests.

vona-ben replied:
@andresbersier as discussed, added a threshold ratio of 0.75

jeffswartz commented on README.md:
```diff
@@ -140,10 +140,12 @@ See the /sample subdirectory (and the /sample/README.md file) for a sample app.
 Opera, and Chromium-based versions of Edge (versions 79+).
 It is not supported in non-Chromium-based versions of Edge.
 
+Update: **Firefox** has been removed from the list of supported browsers for the testQuality feature since version **2.6.0** due to missing some important statistics. This could lead to inaccurate results.
```
Also, please rev the version in package.json (and package-lock.json).

jeffswartz replied:
I was able to run the sample app in Firefox, and it seemed to yield reasonable testQuality() results. We should do one of these things:

* Remove the note if Firefox does support testNetwork().
* Document the fact that Firefox may yield inaccurate results (and describe, if possible).
* Have testNetwork() result in an error if run in Firefox.

vona-ben replied:
@jeffswartz I tested the code and I got firefox not supported as expected:
![Screenshot 2023-05-12 at 4 05 45 PM](https://github.com/opentok/opentok-network-test-js/assets/108658664/3faa6e27-14b6-4813-8ead-6a7b7f5e54cb)

jeffswartz replied:
Yes, please ignore my comment about testQuality working in Firefox. I had some installation issues. I do see the error in Firefox.

jeffswartz commented on src/NetworkTest/types/callbacks.ts:
```diff
@@ -1,4 +1,17 @@
+  packetsLost: number;
+  packetsReceived: number;
+}
+ 
```
Linting error (extra space):
```suggestion

```

jeffswartz commented on sample/src/js/connectivity-ui.js:
```diff
@@ -144,15 +144,15 @@ export function graphIntermediateStats(mediaType, stats) {
    'Bitrate over ' + resultCount[mediaType] + 'sec';
   charts[mediaType].setTitle(null, { text: chartTitle});
-  prevBitsReceived[mediaType] = bitsReceived;
+  prevBitsReceived[mediaType] = bitsSent;
```
When I run the sample, there are two charts: audio bitrate stability and video bitrate stability. Both yield flat results with a constant 0kBps over 29 seconds. This seems broken.

vona-ben replied:
Please can you double-check again? I re-test that from scratch and it works fine on my side.

jeffswartz replied:
Sorry. Issues on my side.

jeffswartz commented on README.md:
```diff
@@ -140,10 +140,12 @@ See the /sample subdirectory (and the /sample/README.md file) for a sample app.
-The `OTNetworkTest.testQuality()` method is supported in Chrome, Firefox, Safari,
+The `OTNetworkTest.testQuality()` method is supported in Chrome, Safari,
 Opera, and Chromium-based versions of Edge (versions 79+).
 It is not supported in non-Chromium-based versions of Edge.
```
```suggestion
It is *not* supported in Firefox and non-Chromium-based versions of Edge.
```
